internal/mesh/internal/controllers/sidecarproxy/builder/service_port_info.go

```diff
@@ -0,0 +1,121 @@
+package builder
+
+import (
+	pbcatalog "github.com/hashicorp/consul/proto-public/pbcatalog/v1alpha1"
+)
+
+// servicePortInfo is a struct used by the destination builder so that it can pre-process
+// what is the service mesh port and what are the distinct ports across the endpoints.
+// This pre-processing reduces the iterations during the BuildDestinations processing
+// would need to get the ports, check if one was the mesh port, know whether it already
+// has a cluster created for it and so on.
+
+type servicePortInfo struct {
+	// meshPortName is the name of the port with Mesh Protocol.
+	meshPortName string
+	// meshPort is the port with Mesh Protocol.
+	meshPort *pbcatalog.WorkloadPort
+	// servicePorts are the distinct ports that need clusters and routers for given destination.
+	// servicePorts do not include the mesh port and are called servicePorts because they
+	// belong to the service.
+	servicePorts map[string]*pbcatalog.WorkloadPort
+}
+
+// newServicePortInfo builds a servicePointInfo given a serviceEndpoints struct. It pre-process
+// what the service mesh port and the distinct service ports across the endpoints for a destination.
+// The following occurs during pre-processing:
+//   - a port must be exposed to at least one workload address on every workload in
+//     the service to be a service port.  Otherwise, the system would risk errors.
+//   - a Workload can optionally define ports specific to workload address.  If no
+//     ports are specified for a workload address, then all the destination ports are
+//     used.
+func newServicePortInfo(serviceEndpoints *pbcatalog.ServiceEndpoints) *servicePortInfo {
+	spInfo := &servicePortInfo{
+		servicePorts: make(map[string]*pbcatalog.WorkloadPort),
+	}
+	type seenData struct {
+		port   *pbcatalog.WorkloadPort
+		seenBy []*int
+	}
+	seen := make(map[string]*seenData)
+	for epIdx, ep := range serviceEndpoints.GetEndpoints() {
+		for _, address := range ep.Addresses {
```

Comment:
Shouldn't it be sufficient to just iterate through all port maps to get the union of all ports? I know we talked about it, but I can't remember now why we need to check if the port is used by the address.

Reply:
Will remove!

Reply:
removed the address checking in e1392e0c628b63537ab0b5ca3b5a25be7e54fc58